Update Daml Documentation to reflect command deduplication related changes [KVL-1094] (#10852)
digital-asset/daml: docs/source/app-dev/services.rst

The on-ledger effect of the command execution will be reported via the `transaction service <#transaction-service>`__, described below. The completion status of the command is reported via the `command completion service <#command-completion-service>`__. Your application should receive completions, correlate them with command submission, and handle errors and failed commands. Alternatively, you can use the `command service <#command-service>`__, which conveniently wraps the command submission and completion services.

> Each intended ledger change is identified by the **change ID**: the triplet of the submitting :ref:`party <com.daml.ledger.api.v1.Commands.party>`, :ref:`application ID <com.daml.ledger.api.v1.Commands.application_id>` and :ref:`command ID <com.daml.ledger.api.v1.Commands.command_id>`.

Reviewer: ```suggestion
Each intended ledger change is identified by the **change ID**, consisting of the submitting :ref:`party <com.daml.ledger.api.v1.Commands.party>`, the :ref:`application ID <com.daml.ledger.api.v1.Commands.application_id>` and the :ref:`command ID <com.daml.ledger.api.v1.Commands.command_id>`.
```

> Each intended ledger change is identified by the **change ID**: the triplet of the submitting :ref:`party <com.daml.ledger.api.v1.Commands.party>`, :ref:`application ID <com.daml.ledger.api.v1.Commands.application_id>` and :ref:`command ID <com.daml.ledger.api.v1.Commands.command_id>`.

Reviewer 2: We take `actAs union party` as the submitting parties. Do you want to change this here, too?

Reviewer: Is saying "the submitting parties" too vague?

Reviewer 2: I find that the links to the corresponding protobuf fields help the reader. It is weird if two fields are linked and we just leave "submitting parties" unanchored.

Reviewer: 8b7eec5, let me know if it's clear or if you think that wording it directly as `actAs union party` with links is better.

Reviewer 2: That's great. Thanks.

> Commands can be labeled with application-specific IDs, all of which are returned in completion events:

Reviewer: ```suggestion
The following application-specific IDs , all of which are included in completion events, can be set in commands:
```

Reviewer: 8b7eec5

> - A :ref:`commandId <com.daml.ledger.api.v1.Commands.command_id>`, returned to the submitting application only. It is generally used to implement this correlation between commands and completions. 

Reviewer: ```suggestion
- A :ref:`commandId <com.daml.ledger.api.v1.Commands.command_id>`, returned to the submitting application only; it can be used to correlate commands to completions.
```
I'd perhaps move it after the submission ID, which seems more important to me.

Reviewer: 8b7eec5

Command deduplication
---------------------

The command submission service deduplicates submitted commands based on their change ID:

- Applications can provide a :ref:`deduplication duration <com.daml.ledger.api.v1.Commands.deduplication_duration>` for each command. If this parameter is not set, the default maximum deduplication period is used.

> - A command submission is considered a duplicate submission if the ledger server receives the command within the deduplication period of a previous command with the same application ID and command ID from the same submitting party.

Reviewer: ```suggestion
- A command submission is considered a duplicate submission if the ledger server receives the command within the deduplication period of a previous command with the same change ID.
```

> - A command submission is considered a duplicate submission if the ledger server receives the command within the deduplication period of a previous command with the same application ID and command ID from the same submitting party.

Reviewer 2: That's not really what's happening in Canton any more. Canton uses the new idea of backward-looking dedup, i.e., the dedup period of the later submission counts, not the one of the accepted submission. I'm not sure what ledger implementations currently still use the forward-looking semantics; kvutils has been changed so that it always uses the max dedup time, which makes the two indistinguishable.

Reviewer: > I'm not sure what ledger implementations currently still use the forward-looking semantics

I'm not sure either, perhaps @bame-da knows more. In addition to that, I'm not sure if the end state is having the backward-looking semantics specified as part of the API, of if the precise behavior should be left unspecified at the API level.

Reviewer 2: > the end state is having the backward-looking semantics specified as part of the API

Yes. That's the desired end state once we have all officially supported ledger implementations moved over.

Reviewer: How about a more general wording until we ensure that all ledgers comply with the new semantics:

```
- A command submission is considered a duplicate submission if the ledger API server is aware of another command within the deduplication period and with the same :ref`change ID <change-id>`.
 ```

Or, we could expand on the new semantics; I think you @bame-da can probably help deciding here.

Reviewer 2: I like your unspecific wording here. It's always fine to make stronger promises in later releases.

Reviewer: Alright, 4792a0a34be0a857ad7285ffda1d5d950a8c11a3. @bame-da (or someone else from product): I'll also wait for your feedback/stamp before merging.
